projects/rocprofiler-sdk/source/include/rocprofiler-sdk/experimental/spm.h

```diff
@@ -0,0 +1,254 @@
+
+/**
+ * @brief (experimental) SPM record flags.
+ *
+ **/
+typedef enum ROCPROFILER_SDK_EXPERIMENTAL rocprofiler_spm_record_flag_t
+{
+    ROCPROFILER_SPM_RECORD_FLAG_DATA_LOST = 0,  ///< records with data loss
+    ROCPROFILER_SPM_RECORD_FLAG_DATA,           ///< records with data
+    ROCPROFILER_SPM_RECORD_FLAG_END,            ///< End of agent service
+    ROCPROFILER_SPM_RECORD_FLAG_LAST,
+} rocprofiler_spm_record_flag_t;
+
+/**
+ * @brief (experimental) Kernel dispatch data for profile counting callbacks.
+ *
+ */
+typedef struct ROCPROFILER_SDK_EXPERIMENTAL rocprofiler_spm_dispatch_counting_service_data_t
+{
+    uint64_t                           size;            ///< Size of this struct
+    rocprofiler_async_correlation_id_t correlation_id;  ///< Correlation ID for this dispatch
+    rocprofiler_kernel_dispatch_info_t dispatch_info;   ///< Dispatch info
+} rocprofiler_spm_dispatch_counting_service_data_t;
+
+/**
+ * @brief (experimental) SPM record counter record.
+ *
+ **/
+typedef struct ROCPROFILER_SDK_EXPERIMENTAL rocprofiler_spm_counter_record_t
+{
+    uint64_t size;  ///< Size of this structure. Used for versioning and validation.
+    rocprofiler_counter_instance_id_t id;         ///< Counter instance id
+    rocprofiler_agent_id_t            agent_id;   ///< Agent on which the record is collected
+    rocprofiler_timestamp_t           timestamp;  ///< timestamp of the sample
+    uint64_t value;  ///< SPM sample for the counter with counter instance id: id
+} rocprofiler_spm_counter_record_t;
+
+/**
+ * @brief (experimental) Callback to receive SPM data
+ *
+ * @param [in] dispatch_data kernel dispatch data
+ * @param [in] records pointer to the array of SPM records
+ * @param [in] record_count  size of the record array
+ * @param [in] flags  rocprofiler_spm_record_flag_t
+ * @param [in] userdata user data supplied by dispatch callback
+ * @param [in] record Callback data supplied via dispatch configure service
+ */
+ROCPROFILER_SDK_EXPERIMENTAL
+typedef void (*rocprofiler_spm_dispatch_counting_record_cb_t)(
+    rocprofiler_spm_dispatch_counting_service_data_t dispatch_data,
+    rocprofiler_spm_counter_record_t*                records,
+    size_t                                           record_count,
+    uint8_t                                          flags,
+    rocprofiler_user_data_t*                         userdata,
+    void*                                            record_callback_args);
+/**
+ * @brief (experimental) Callback query if dispatch should be profiled
+ *
+ * @param [in] dispatch_data kernel dispatch data
+ * @param [in] config  spm counter config
+ * @param [out] user_data User data unique to this dispatch. Returned in record callback
+ * @param [in] callback_data_args Callback supplied via dispatch configure service
+ */
+ROCPROFILER_SDK_EXPERIMENTAL
+typedef void (*rocprofiler_spm_dispatch_counting_service_cb_t)(
+    rocprofiler_spm_dispatch_counting_service_data_t dispatch_data,
```

Comment:
```suggestion
    const rocprofiler_spm_dispatch_counting_service_data_t* dispatch_data,
```
The `rocprofiler_spm_dispatch_counting_service_data_t` struct has a `size` field, indicating it may be extended in the future and thus we need to pass this via pointer for ABI reasons